ENIGMAsystem/SHELL/Platforms/General/PFwindow.h

```diff
@@ -156,8 +157,9 @@ enum {
 
 extern double mouse_x, mouse_y;
 extern int mouse_button, mouse_lastbutton;
-extern std::string keyboard_lastchar;
 extern int keyboard_lastkey;
+extern std::string keyboard_string;
+extern std::string keyboard_lastchar;
```

Comment:
why not char?

Reply:
Ask Mark Overmars, I don't know.

Reply:
dafux it got do with mark? last_char stores string or char?

Reply:
Oh, because it supports UTF-8 in GMSv1.4 and ENIGMA it must be string. Also the old GM8.1 and lower manual states that it is of type string, but there was no char type in old GM.

Reply:
I dont think you need a variable at all as you can just return a sunstring of first var no?

Reply:
You need a separate variable because of tkg's issue actually. The lastchar recognizes tab and enter key, but it should not be appended to the `keyboard_string` variable. That's the very thing I just got stuck on trying to figure out if Win32 is handling that or GM handled it specially itself in WM_CHAR so that our SDL platform is consistent.
https://github.com/enigma-dev/enigma-dev/issues/2080#issuecomment-675753715